### DISCUSSION TITLE
v7 新規作成ボタンのモーダルをデフォルトにする設定 / Config option setting the modal new page window as the default

### DISCUSSION THREAD

**Discussion by nandnor93** [growilabs/growi#8870] [Ideas / Requests] [answered]
Slackで発言させていただいた件です。
https://wsgrowi.slack.com/archives/C01U5NBHR5F/p1717147938766469

## 趣旨/ Overview

- v6までの新規作成ボタンで表示されていたページ作成モーダルは、v7ではサブメニュー項目になり、作成ボタンのデフォルトは閲覧中ページの子ページをWIPで作成する挙動になった。  
    The new page modal window, which was the only behavior when we clicked the New Page button in v6, has become a submenu item, and the default has become to create a WIP child page just under the currently displayed page.
- v6までの使用方法に慣れているユーザがそのまま新規作成ボタンを押すことにより、**意図しないWIPページがたくさん生成される**事態となった。  
    The users who have used v6 for a while click the New Page button, resulting in **a considerable number of unintended "WIP" pages**.
- ページ作成モーダルがサブメニュー項目として存在している以上、これを再度デフォルトに据えることについて問題は起こり得ないはずである。  
    Meanwhile, the New Page Modal does exist in the current implementation as a submenu item. Thus, bringing back it default would not pose any potential issues.
- デフォルトをv7の新規WIP子ページにするか、従来のモーダルにするかを選択する**設定オプションを追加**してはいかがだろうか。  
    I think it would be useful if we had a **configuration option to choose the behavior of the New Page button** from the v7 style (creating a child page immediately) or the conventional style (displaying the New Page Modal).

続きは後ほど書き足します。まずは趣旨にて失礼します。

**Comment by nandnor93** [comment]
## 暫定対策 / Workaround 

当方の環境ではソースを改変し、新規作成ボタンのイベントハンドラをサブメニュー項目のそれに差し替えることによりv6動作を復元してDockerをビルドし、使用しています。  
つまり、設定項目でこれを差し替えられるようにしてはいかがかという提案です。

In our environment, I replaced the event handler of the New Page button with that of the New Page Modal submenu item in the source and manually built the Docker image to revert to the v6 behavior.  
This is possible; so I am proposing adding a configuration option to do so.

https://github.com/weseek/growi/blob/ff51aa827e4662e864b7a8fd27aa2e90ec075ec5/apps/app/src/components/Sidebar/PageCreateButton/PageCreateButton.tsx#L58

```typescript:growi/apps/app/src/components/Sidebar/PageCreateButton/PageCreateButton.tsx
export const PageCreateButton = React.memo((): JSX.Element => {

  // ........

  return (
    <div
      className="d-flex flex-row mt-2"
    >
      <div className="btn-group flex-grow-1">
        <CreateButton
          className="z-2"
          onClick={() => openPageCreateModal(currentPagePath)}   //   <---- {createNewPageWithToastr}
          disabled={isNewPageCreating || isTodaysPageCreating || isTemplatePageCreating}
        />

        // ........

  );
});

```

**Comment by nandnor93** [comment]
また、別件をつなげるのは恐縮ですが、そのページ作成モーダルの挙動もv6から変わっていることに気づきました。  
モーダルのユーザページのパスが「メモ」に固定になったことは意図がありますでしょうか？  
私の使い方では `/user/nandnor93/private` というパスを作って限定公開ページにしてあり、パスの「メモ」を「private」に書き換えたら自動的に限定公開で新規作成ができる、という使い方をv6までしていたのですが、v7にしたらこれができなくなりました。

↓v6のモーダル  
<img width="809" alt="image" src="https://github.com/weseek/growi/assets/67882711/579a10d1-786f-4eff-8a56-2691eb241961">

↓v7のモーダル  
<img width="890" alt="image" src="https://github.com/weseek/growi/assets/67882711/2d642ab8-c787-466b-a297-efe5866b6e20">

**Comment by nandnor93** [comment]
当方のユースケースとしては

- まずトップを開く
- とりあえず新規作成ボタンを押す
- 日付順ユーザページを書く

が主な手順だったので、これのつもりで新規作成ボタンを押す習慣のあるユーザがたくさんいて、トップ以下にWIPがたくさんできてしまいました。

**Comment by yuki-takei** [comment, marked as answer]
完全な回答ではないかもしれませんが、v7 の UI の設計思想と現在の仕様についてアウトプットします。

## ページ作成アクションの数
まず v7 でのページ作成はアクション数を減らす方向の模索をしています。
従来のモーダルを一度開いて作成ではなく、ボタンワンクリックでその時のコンテキストに合わせてページを作成するような設計になっています。例えば「今閲覧中のページの配下に Untitled-1 という子ページを作る」という具合です。

これ自体は昨今のナレッジベースの標準の動作という認識で、なぜトレンドがこうなるのかというと「まず本文を書き始めたい」という利用者を想定しているのではないかという推測です。

## WIP ページの取り扱いについて

> v6までの使用方法に慣れているユーザがそのまま新規作成ボタンを押すことにより、意図しないWIPページがたくさん生成される事態となった。

この事象は予期されていたので、v7 では [WIP ページの自動削除機能が搭載されています](https://docs.growi.org/ja/admin-guide/upgrading/70x.html#%E5%88%A9%E7%94%A8%E8%80%85%E5%90%91%E3%81%91)。また、サイドバーのページツリーや最近の更新リスト中でも WIP ページを表示しない設定を選択できるようになっています。

ですので、v7 は利用側に対しては「WIP は量産されるかもしれないけれど気にしないで済む運用」がなされることを期待しています。

### WIP ページ溢れをどう捉えるか

僕自身も Wiki 管理者として Untitled-1 のような WIP ページがたくさんできることを気持ち悪いと思う側ですが、Wiki 運用にあたっては**利用者の感覚と管理者の感覚は分けて捉える必要がある**と思います。

GROWI にご意見くださる方は得てして管理者の方が多いのでやはり情報の整理整頓にこだわりのある方が多いのですが、実際利用者は整頓にそこまで頓着がないこともままあります。一方で管理者の課題としては「文書を残してくれないその他大勢にどうナレッジベース活用を習慣付けるか」がどんな組織でも重要で、そのため開発者として意見するなら、「まず本文を書き始めたい」場合によっては「すぐにでも書き殴りたい」ような利用者が GROWI の使い始めで躓かないような使用感にシフトしたいのはあり、もしページ作成までのアクションを減らす UI や挙動が効果的なのであれば管理者サイドはそのギャップや散らかることとのトレードオフを認識しつつどう埋めていくのかを考えることもまた重要だろうという意見です。

## v7 の評価と今後の改善について

ただ v7 以降から使い始めたユーザーさんがどういう感覚でいるのか、v7 は 本当に v6 よりも利用者にウケる UI・挙動になっているのかはまだわからないので、時間が必要かなと思っています。また旧来のバージョンからのユーザーさんたちが時間経過と共にその仕様に馴染めるのかも同様に今後の経過次第かなと考えており、従ってページ作成ボタンの仕様変更や動作変更オプションの提供ももう少し様子を見たいところです。

**Reply by yuki-takei** [reply]
> また、別件をつなげるのは恐縮ですが、そのページ作成モーダルの挙動もv6から変わっていることに気づきました。
> モーダルのユーザページのパスが「メモ」に固定になったことは意図がありますでしょうか？

これも「アクション数を減らす」方向性に対する変更の一環で、モーダルを開かない場合にそこの「メモ」部分の書き換え UI はかなり難しいのですよね。書き換えて新規作成している人がいるかどうかを弊社内で調査したのですがほぼゼロだったため不要と判断しました。(狭い文化圏でのサンプルで申し訳ないです)

モーダルを開いた時には書き換えられるように復活させるくらいはアリかもしれないです。

**Comment by nandnor93** [comment]
ご返信ありがとうございます。    
当方のユースケースではひとまず上述のソース改変で対応しておりますので、ひとまずの困り事は解決しております。


その上で、

> 「まず本文を書き始めたい」という利用者

これはわかる気がする一方、従来の使用方法に慣れているとかえってストレスでしかないと感じました。  
その意味では、「ユーザごとに」これを選べて、新規ユーザのデフォルトはv7形式にする、あたりが妥当に思いますが、実装・保守の手間とのトレードオフというのは理解します。

一番困ったのが、プロジェクトごとにグループを設定した限定公開ページが分かれているのですが、これの作成場所を誤ると権限移動が面倒なので削除再作成になることでした。  
先に挙げた `/user/hoge/private/...` もそれに類似します。

> 文書を残してくれないその他大勢にどうナレッジベース活用を習慣付けるか

これは本当にそうで、当方でも頻繁に問題にあがります。  
ユーザメモページの推奨はその延長上にあります。  
v7ではプルダウンメニューからユーザメモを直接作成できるようになったのは良いですね。（ただし、それが末端ユーザに普及するかは別問題なのが悩ましいところです……）

----


> WIP ページを表示しない設定

こちら見つけられなかったのですが、v7.0.8以降の機能でしょうか……？  
少なくとも私の使い方では、時間的局所性に頼った運用をしているとでもいいましょうか、直近の会議の議事録につなげて清書ページを作ることが多いので、WIPで流れてゆかないように設定できるならそうしたいです。



> WIP ページの自動削除機能

これは自動h1挿入と一緒に使用可能でしょうか？うまく消えてくれなかった気がするのは、待ち足りなかったのかもしれませんが。




----

当方も本業の片手間につき、書き殴りにて失礼いたしました。  
まさにこれは利用者全体の利便を見るべきと思いますので、皆様の意見を待ちたいと思います。

**Reply by yuki-takei** [reply]
>> WIP ページを表示しない設定
>
> こちら見つけられなかったのですが、v7.0.8以降の機能でしょうか……？

どこからかは覚えていないですが、
システムグローバルなせっていではなくユーザーが各々で変更可能なサイドバーの設定です。
Page Tree と Recent Changes のコンテンツの右上のメニューから変更できます。

>> WIP ページの自動削除機能
> 
> これは自動h1挿入と一緒に使用可能でしょうか？

たぶん大丈夫だと思いますが、うまく消えないケースがあるのかどうかちょっと自信がないです。
